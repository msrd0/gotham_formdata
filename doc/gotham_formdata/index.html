<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This crate is an extension to the popular gotham web framework for Rust. It aims to reduce boilerplate necessary to read request bodies today as a stop-gap until gotham finally implements a body extractor."><meta name="keywords" content="rust, rustlang, rust-lang, gotham_formdata"><title>gotham_formdata - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../gotham_formdata/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><p class="location">Crate gotham_formdata</p><div class="block version"><p>Version 0.0.4</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all gotham_formdata's items</p></a><div class="block items"><ul><li><a href="#modules">Modules</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div><p class="location"></p><div id="sidebar-vars" data-name="gotham_formdata" data-ty="mod" data-relpath="../"></div></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" class="help-button">?</button>
                <a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="">gotham_formdata</a></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/gotham_formdata/lib.rs.html#1-105" title="goto source code">[src]</a></span></h1><div class="docblock"><p>This crate is an extension to the popular <a href="https://github.com/gotham-rs/gotham">gotham web framework</a> for Rust. It aims to
reduce boilerplate necessary to read request bodies today as a stop-gap until gotham finally
implements a <a href="https://github.com/gotham-rs/gotham/issues/11">body extractor</a>.</p>
<h1 id="sparkles-features" class="section-header"><a href="#sparkles-features">:sparkles: Features</a></h1>
<ul>
<li>Parse <code>application/x-www-form-urlencoded</code> request bodies</li>
<li>Parse <code>multipart/form-data</code> request bodies</li>
<li>Verify the parsed request body</li>
<li><code>#![forbid(unsafe_code)]</code> ensures that all functionality is implemented in 100% safe Rust code</li>
</ul>
<h1 id="warning-warning" class="section-header"><a href="#warning-warning">:warning: Warning</a></h1>
<p>This crate is asynchronous, but does not yet enforce uploads limits. <strong>YOU ARE RESPONSIBLE
FOR ENFORCING UPLOAD LIMITS.</strong></p>
<h1 id="spiral_notepad-example" class="section-header"><a href="#spiral_notepad-example">:spiral_notepad: Example</a></h1>
<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">gotham_formdata</span>::<span class="ident">FormData</span>;

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">FormData</span>)]</span>
<span class="kw">struct</span> <span class="ident">LoginData</span> {
	<span class="attribute">#[<span class="ident">validate</span>(<span class="ident">regex</span> <span class="op">=</span> <span class="string">&quot;[a-zA-Z0-9_]&quot;</span>, <span class="ident">error</span> <span class="op">=</span> <span class="string">&quot;The username contains illegal characters.&quot;</span>)]</span>
	<span class="ident">username</span>: <span class="ident">String</span>,
	<span class="attribute">#[<span class="ident">validate</span>(<span class="ident">min_length</span> <span class="op">=</span> <span class="number">6</span>)]</span>
	<span class="ident">password</span>: <span class="ident">String</span>
}

<span class="kw">async</span> <span class="kw">fn</span> <span class="ident">login_handler</span>(<span class="ident">state</span>: <span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="ident">State</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span><span class="ident">Response</span><span class="op">&lt;</span><span class="ident">Body</span><span class="op">&gt;</span>, <span class="ident">HandlerError</span><span class="op">&gt;</span> {
	<span class="kw">let</span> <span class="ident">login_data</span>: <span class="ident">LoginData</span> <span class="op">=</span> <span class="ident">FormData</span>::<span class="ident">parse_form_data</span>(<span class="ident">state</span>).<span class="kw">await</span><span class="question-mark">?</span>;
	<span class="prelude-val">Ok</span>(<span class="kw">if</span> <span class="ident">login_data</span>.<span class="ident">password</span> <span class="op">=</span><span class="op">=</span> <span class="string">&quot;secret&quot;</span> {
		<span class="ident">create_response</span>(<span class="ident">state</span>, <span class="ident">StatusCode</span>::<span class="ident">OK</span>, <span class="ident">TEXT_PLAIN</span>, <span class="ident">login_data</span>.<span class="ident">username</span>)
	} <span class="kw">else</span> {
		<span class="ident">create_empty_response</span>(<span class="ident">state</span>, <span class="ident">StatusCode</span>::<span class="ident">FORBIDDEN</span>)
	})
}</pre></div>
<h1 id="label-versioning" class="section-header"><a href="#label-versioning">:label: Versioning</a></h1>
<p>Like all rust crates, this crate will follow semantic versioning guidelines. However, changing
the MSRV (minimum supported rust version) is not considered a breaking change.</p>
<h1 id="page_with_curl-license" class="section-header"><a href="#page_with_curl-license">:page_with_curl: License</a></h1>
<pre><code class="language-text">Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
</code></pre>
</div><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<table><tr class="module-item"><td><a class="mod" href="conversion/index.html" title="gotham_formdata::conversion mod">conversion</a></td><td class="docblock-short"><p>This mod contains conversion traits for common used types, that allows them to be created from
a stream of bytes. Furthermore, it allows every type that implements <a href="https://doc.rust-lang.org/nightly/core/str/traits/trait.FromStr.html" title="FromStr">FromStr</a> plus some other
common types to be converted.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="validate/index.html" title="gotham_formdata::validate mod">validate</a></td><td class="docblock-short"><p>This mod contains the <a href="../gotham_formdata/validate/trait.Validator.html" title="Validator">Validator</a> trait as well as pre-defined validation methods.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="value/index.html" title="gotham_formdata::value mod">value</a></td><td class="docblock-short"><p>This mod defines the <a href="../gotham_formdata/value/struct.Value.html" title="Value">Value</a> used for parsing the form data.</p>
</td></tr></table><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<table><tr class="module-item"><td><a class="enum" href="enum.Error.html" title="gotham_formdata::Error enum">Error</a></td><td class="docblock-short"><p>This error type is used when parsing form data from a request body was unsuccessful.</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.FormData.html" title="gotham_formdata::FormData trait">FormData</a></td><td class="docblock-short"><p>This is the trait implemented by <code>#[derive(FormData)]</code>. It provides a method to parse the struct
it is implemented for to be parsed from the request body contained in gotham’s state.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.FormDataFromState.html" title="gotham_formdata::FormDataFromState trait">FormDataFromState</a></td><td class="docblock-short"><p>This is the equivalent of <a href="../gotham_formdata/trait.FormData.html" title="FormData">FormData</a> from the state’s perspective. Use this if you prefer
<code>state.parse_form_data::&lt;MyData&gt;()?</code> over <code>MyData::parse_form_data(&amp;mut state)?</code>.</p>
</td></tr></table><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.FormDataFuture.html" title="gotham_formdata::FormDataFuture type">FormDataFuture</a></td><td class="docblock-short"><p>This is the return type of <a href="../gotham_formdata/trait.FormData.html#tymethod.parse_form_data" title="FormData::parse_form_data">FormData::parse_form_data</a>.</p>
</td></tr></table><h2 id="derives" class="section-header"><a href="#derives">Derive Macros</a></h2>
<table><tr class="module-item"><td><a class="derive" href="derive.FormData.html" title="gotham_formdata::FormData derive">FormData</a></td><td class="docblock-short"><p>This derive macro implements <code>FormData</code> for the struct it is invoked on. Enums, unions and
tuple structs are not supported.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="gotham_formdata" data-search-js="../search-index.js"></div>
    <script src="../main.js"></script></body></html>