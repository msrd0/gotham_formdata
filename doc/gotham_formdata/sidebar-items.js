initSidebarItems({"derive":[["FormData","This derive macro implements `FormData` for the struct it is invoked on. Enums, unions and tuple structs are not supported."]],"enum":[["Error","This error type is used when parsing form data from a request body was unsuccessful."]],"mod":[["conversion","This mod contains conversion traits for common used types, that allows them to be created from a stream of bytes. Furthermore, it allows every type that implements [FromStr] or [From<&[u8]>] to be converted."],["validate","This mod contains the [Validator] trait as well as pre-defined validation methods."]],"trait":[["FormData","This is the trait implemented by `#[derive(FormData)]`. It provides a method to parse the struct it is implemented for to be parsed from the request body contained in gotham's state."],["FormDataFromState","This is the equivalent of [FormData] from the state's perspective. Use this if you prefer `state.parse_form_data::<MyData>()?` over `MyData::parse_form_data(&mut state)?`."]],"type":[["FormDataFuture","This is the return type of [FormData::parse_form_data]."]]});