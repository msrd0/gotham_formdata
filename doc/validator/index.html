<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Example:"><meta name="keywords" content="rust, rustlang, rust-lang, validator"><title>validator - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-1f7d512b176f0f72.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-124a1ca42af929b6.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-93196c7a1c3542a8.css" id="mainThemeStyle"><link rel="stylesheet" id="themeStyle" href="../static.files/light-4743e13df3dfe8c4.css"><link rel="stylesheet" disabled href="../static.files/dark-0e1b889528bd466b.css"><link rel="stylesheet" disabled href="../static.files/ayu-65289d5d067c7c66.css"><script id="default-settings" ></script><script src="../static.files/storage-d43fa987303ecbbb.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-3367e395607fafc1.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../validator/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../validator/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate validator</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.16.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-5ec35bf9ca753509.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">validator</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/validator/lib.rs.html#1-93">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="example"><a href="#example">Example:</a></h2>
<div class="example-wrap ignore"><div class='tooltip'>ⓘ</div><pre class="rust rust-example-rendered"><code><span class="kw">use </span>serde::Deserialize;

<span class="comment">// A trait that the Validate derive will impl
</span><span class="kw">use </span>validator::{Validate, ValidationError};

<span class="attr">#[derive(Debug, Validate, Deserialize)]
</span><span class="kw">struct </span>SignupData {
    <span class="attr">#[validate(email)]
    </span>mail: String,
    <span class="attr">#[validate(phone)]
    </span>phone: String,
    <span class="attr">#[validate(url)]
    </span>site: String,
    <span class="attr">#[validate(length(min = <span class="number">1</span>), custom = <span class="string">&quot;validate_unique_username&quot;</span>)]
    #[serde(rename = <span class="string">&quot;firstName&quot;</span>)]
    </span>first_name: String,
    <span class="attr">#[validate(range(min = <span class="number">18</span>, max = <span class="number">20</span>))]
    </span>age: u32,
}

<span class="kw">fn </span>validate_unique_username(username: <span class="kw-2">&amp;</span>str) -&gt; <span class="prelude-ty">Result</span>&lt;(), ValidationError&gt; {
    <span class="kw">if </span>username == <span class="string">&quot;xXxShad0wxXx&quot; </span>{
        <span class="comment">// the value of the username will automatically be added later
        </span><span class="kw">return </span><span class="prelude-val">Err</span>(ValidationError::new(<span class="string">&quot;terrible_username&quot;</span>));
    }

    <span class="prelude-val">Ok</span>(())
}

<span class="kw">match </span>signup_data.validate() {
  <span class="prelude-val">Ok</span>(<span class="kw">_</span>) =&gt; (),
  <span class="prelude-val">Err</span>(e) =&gt; <span class="kw">return </span>e;
};</code></pre></div>
<h2 id="available-validations"><a href="#available-validations">Available Validations:</a></h2><div><table><thead><tr><th>Validation</th><th>Notes</th></tr></thead><tbody>
<tr><td><code>email</code></td><td></td></tr>
<tr><td><code>url</code></td><td></td></tr>
<tr><td><code>length</code></td><td></td></tr>
<tr><td><code>range</code></td><td></td></tr>
<tr><td><code>must_match</code></td><td></td></tr>
<tr><td><code>contains</code></td><td></td></tr>
<tr><td><code>does_not_contain</code></td><td></td></tr>
<tr><td><code>custom</code></td><td></td></tr>
<tr><td><code>regex</code></td><td></td></tr>
<tr><td><code>credit_card</code></td><td>(Requires the feature <code>card</code> to be enabled)</td></tr>
<tr><td><code>phone</code></td><td>(Requires the feature <code>phone</code> to be enabled)</td></tr>
<tr><td><code>non_control_character</code></td><td>(Required the feature <code>unic</code> to be enabled)</td></tr>
<tr><td><code>nested</code></td><td>(Uses the validation of the field type it self)</td></tr>
<tr><td><code>required</code></td><td></td></tr>
</tbody></table>
</div>
<p><a href="https://github.com/Keats/validator/blob/master/README.md">Checkout the project README of an in-depth usage description with examples.</a></p>
<h2 id="installation"><a href="#installation">Installation:</a></h2>
<p>Add the validator to the dependencies in the Cargo.toml file.</p>
<div class="example-wrap"><pre class="language-toml"><code>[dependencies]
validator = { version = &quot;0.12&quot;, features = [&quot;derive&quot;] }
</code></pre></div></div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.ValidationError.html" title="validator::ValidationError struct">ValidationError</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.ValidationErrors.html" title="validator::ValidationErrors struct">ValidationErrors</a></div></div></div><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ValidationErrorsKind.html" title="validator::ValidationErrorsKind enum">ValidationErrorsKind</a></div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Contains.html" title="validator::Contains trait">Contains</a></div><div class="item-right docblock-short">Trait to implement if one wants to make the <code>contains</code> validator
work for more types</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.HasLen.html" title="validator::HasLen trait">HasLen</a></div><div class="item-right docblock-short">Trait to implement if one wants to make the <code>length</code> validator
work for more types</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Validate.html" title="validator::Validate trait">Validate</a></div><div class="item-right docblock-short">This is the original trait that was implemented by deriving <code>Validate</code>. It will still be
implemented for struct validations that don’t take custom arguments. The call is being
forwarded to the <code>ValidateArgs&lt;'v_a&gt;</code> trait.</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.ValidateArgs.html" title="validator::ValidateArgs trait">ValidateArgs</a></div><div class="item-right docblock-short">This trait will be implemented by deriving <code>Validate</code>. This implementation can take one
argument and pass this on to custom validators. The default <code>Args</code> type will be <code>()</code> if
there is no custom validation with defined arguments.</div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_contains.html" title="validator::validate_contains fn">validate_contains</a></div><div class="item-right docblock-short">Validates whether the value contains the needle
The value needs to implement the Contains trait, which is implement on String, str and Hashmap<String>
by default.</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_does_not_contain.html" title="validator::validate_does_not_contain fn">validate_does_not_contain</a></div><div class="item-right docblock-short">Validates whether the value does not contain the needle
The value needs to implement the Contains trait, which is implement on String, str and Hashmap<String>
by default.</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_email.html" title="validator::validate_email fn">validate_email</a></div><div class="item-right docblock-short">Validates whether the given string is an email based on the <a href="https://html.spec.whatwg.org/multipage/forms.html#valid-e-mail-address">HTML5 spec</a>.
<a href="https://tools.ietf.org/html/rfc5322">RFC 5322</a> is not practical in most circumstances and allows email addresses
that are unfamiliar to most users.</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_ip.html" title="validator::validate_ip fn">validate_ip</a></div><div class="item-right docblock-short">Validates whether the given string is an IP</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_ip_v4.html" title="validator::validate_ip_v4 fn">validate_ip_v4</a></div><div class="item-right docblock-short">Validates whether the given string is an IP V4</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_ip_v6.html" title="validator::validate_ip_v6 fn">validate_ip_v6</a></div><div class="item-right docblock-short">Validates whether the given string is an IP V6</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_length.html" title="validator::validate_length fn">validate_length</a></div><div class="item-right docblock-short">Validates the length of the value given.
If the validator has <code>equal</code> set, it will ignore any <code>min</code> and <code>max</code> value.</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_must_match.html" title="validator::validate_must_match fn">validate_must_match</a></div><div class="item-right docblock-short">Validates that the 2 given fields match.
Both fields are optionals</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_range.html" title="validator::validate_range fn">validate_range</a></div><div class="item-right docblock-short">Validates that the given <code>value</code> is inside the defined range. The <code>max</code> and <code>min</code> parameters are
optional and will only be validated if they are not <code>None</code></div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_required.html" title="validator::validate_required fn">validate_required</a></div><div class="item-right docblock-short">Validates whether the given Option is Some</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.validate_url.html" title="validator::validate_url fn">validate_url</a></div><div class="item-right docblock-short">Validates whether the string given is a url</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="validator" data-themes="" data-resource-suffix="" data-rustdoc-version="1.68.2 (9eb3afe9e 2023-03-27)" data-search-js="search-98d53477a794af0b.js" data-settings-js="settings-c3c521c753752a1a.js" data-settings-css="settings-08ddfdda51b8ee2e.css" ></div></body></html>