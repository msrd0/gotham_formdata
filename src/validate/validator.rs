use std::convert::Infallible;

/// This trait allows data of type `T` to be verified against custom criteria.
///
/// Example:
///
/// ```rust
/// use gotham_formdata::FormData;
///
/// fn validate_password(password: &String) -> Result<(), &'static str> {
/// 	if password.len() < 8 {
/// 		return Err("Password is too short");
/// 		}
/// 	Ok(())
/// 	}
///
/// #[derive(FormData)]
/// struct LoginData {
/// 	username: String,
/// 	#[validate(validator = validate_password)]
/// 	password: String
/// 	}
/// ```
pub trait Validator<T> {
	/// The error returned when validation failed.
	type Err;

	/// Performs the validation.
	fn validate(self, data: &T) -> Result<(), Self::Err>;
}

/// Convert `()` into an allways accepting validator.
impl<T> Validator<T> for () {
	type Err = Infallible;

	fn validate(self, _: &T) -> Result<(), Infallible> {
		Ok(())
	}
}

/// Convert any function with the correct signature into a validator.
impl<F, Err, T> Validator<T> for F
where
	F: Fn(&T) -> Result<(), Err>
{
	type Err = Err;

	fn validate(self, data: &T) -> Result<(), Err> {
		self(data)
	}
}
