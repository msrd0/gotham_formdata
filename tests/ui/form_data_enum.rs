use gotham_formdata::FormData;
use validator::Validate;

#[derive(FormData, Validate)]
enum MyFormData {}

fn main() {}
