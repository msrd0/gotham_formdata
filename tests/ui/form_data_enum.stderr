error: #[derive(FormData)] can only be used on structs
 --> $DIR/form_data_enum.rs:4:10
  |
4 | #[derive(FormData, Validate)]
  |          ^^^^^^^^
  |
  = note: this error originates in the derive macro `FormData` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[derive(Validate)] can only be used with structs
 --> $DIR/form_data_enum.rs:5:1
  |
5 | enum MyFormData {}
  | ^^^^
